inc/targetticket.class.php

```diff
@@ -816,6 +816,32 @@ public function save(PluginFormcreatorFormAnswer $formanswer) {
          $data['_tag_content'][] = $document['tag'];
       }
 
+   //KKK
+   $saved_documents = $formanswer->getFileProperties();
+
+   if ($saved_documents) {
+      foreach ($formanswer->getFileFields() as $questionId) {
+	 $data["_filename"]=array_merge($data["_filename"],$saved_documents["_filename"][$questionId]);
+	 $data["_tag_filename"]=array_merge($data["_tag_filename"],$saved_documents["_tag_filename"][$questionId]);
+
+	 foreach ( $saved_documents["_filename"][$questionId] as $key => $filename) {
+	    $uploaded_filename=$formanswer->getFileName($questionId,$key);
+	    copy(GLPI_DOC_DIR . '/' . $uploaded_filename, GLPI_TMP_DIR . '/' . $filename);
+	 }
+      }
+
+   } else {
+      foreach ($formanswer->getFileFields() as $questionId) {
+	 $data["_filename"]=array_merge($data["_filename"],$formanswer->input["_formcreator_field_".$questionId]);
+	 $data["_prefix_filename"]=array_merge($data["_prefix_filename"],$formanswer->input["_prefix_formcreator_field_".$questionId]);
+	 $data["_tag_filename"]=array_merge($data["_tag_filename"],$formanswer->input["_tag_formcreator_field_".$questionId]);
+	 foreach ( $formanswer->input["_formcreator_field_".$questionId] as $key => $filename) {
+	    $uploaded_filename=$formanswer->getFileName($questionId,$key);
+	    copy(GLPI_DOC_DIR . '/' . $uploaded_filename, GLPI_TMP_DIR . '/' . $filename);
+	 }
+      }
```

Comment:
Hi

It seems we never enter this part of code.

When Formcreator calls save() method, we always have a form answer available. I did some tests and as I expected I could not enter here. I think it is dead code. Can you tell in which case this part is used ?

Note: This is the last point to solve before finishing my port for Formcretor 2.13. If you have a GLPI 10 for tests, you may want to test the version I prepared.

Reply:
Let me check... 
The first thing I can see is that my getFileProperties function never returns NULL. I have to change this in order to avoid php warnings when no file is attached.
Next, it seems that you are right. As far as I remember, I created the first part of the if statement in order to manage forms requiring validation. It seems that the test is also true for immediate ticket creation as files are already in the DB. But this way we are loosing eventual _prefix_filename field from the input...

Actually, I don't have GLPI 10 instance...

Reply:
wait before changing your PR. Have a look instead on my version for 2.13 : I reworked the code of this method (and it moved in an abstract class).

>
As far as I remember, I created the first part of the if statement in order to manage forms requiring validation. 

I tested this case, and even when the form is under validation we don't enter in the part I pointed. This is normal : the form answer is created before  the validation step.

Then in all cases we loose the _prefix_ and this might be a problem to have the documents in the notifications. Notification is the last thing I must to check on my side

Reply:
With the PR code, I have documents in the notifications as I copy files back to the temp folder. That is probably the reason why missing _prefix_filename is not really an issue... Need confirmation of "prefix" role in the original code...
The else statement is not correct and not needed as if getFileProperties returns an empty array we don't have to fill documents fields.

Function remake of getFileProperties should be better:

   public function getFileProperties() : Array {

      $formcreator_field=[];
      $tag_formcreator_field=[];
      $formcreator_field_array=[];
      $tag_formcreator_field_array=[];

      $form = $this->getForm();

      foreach ($this->getFileFields() as $questionId) {

         foreach ($this->getQuestionFields($form)[$questionId]->getDocumentsForTarget() as $key => $documentId){
            $doc = new Document();
            $doc->getFromDB($documentId);
            $formcreator_field[]= $doc->fields['filename'];
            $tag_formcreator_field[]= $doc->fields['tag'];
         }
         if ($formcreator_field && $tag_formcreator_field) {
            $formcreator_field_array[$questionId] = $formcreator_field;
            $tag_formcreator_field_array[$questionId] = $tag_formcreator_field;
         }
      }

      if ($formcreator_field_array && $tag_formcreator_field_array) {
         return [ "_filename" => $formcreator_field_array,
                  "_tag_filename" => $tag_formcreator_field_array];
      } else {
        return [];
      }
   }

Not sure to have time to check 2.13 today...

Reply:
Hi 

FYI : the version I built from your PR for 2.13 has been merged.